File: gss-ao/backend/ingestion/doc_converter.py
"""Conversion et extraction de texte multi-formats — fondation des parseurs.

Responsabilités :
  - détecter le binaire LibreOffice (`soffice`) — PATH ou emplacement macOS ;
  - convertir un `.doc` legacy en `.docx` via LibreOffice headless ;
  - charger un `.docx` (python-docx) — natif ou issu d'une conversion ;
  - extraire le texte d'un PDF (PyMuPDF) — fallback OCR OPTIONNEL hors périmètre.

Principe de sûreté : si LibreOffice est requis mais absent, on lève
`LibreOfficeNotFoundError` avec un message actionnable — JAMAIS un résultat
silencieusement faux (exigence brief §12 : revue humaine, pas d'illusion).
"""

from __future__ import annotations

import shutil
import subprocess
import tempfile
from pathlib import Path

from backend.core.config import Settings, get_settings

# Emplacements macOS usuels de LibreOffice (hors PATH).
_MACOS_SOFFICE_CANDIDATES = (
    "/Applications/LibreOffice.app/Contents/MacOS/soffice",
    "/opt/homebrew/bin/soffice",
    "/usr/local/bin/soffice",
)


class DocConverterError(RuntimeError):
    """Erreur générique de conversion/extraction."""


class LibreOfficeNotFoundError(DocConverterError):
    """LibreOffice est requis (conversion .doc) mais introuvable."""


def find_soffice(settings: Settings | None = None) -> str | None:
    """Localise le binaire `soffice`.

    Ordre : `SOFFICE_BIN` (config) > PATH (`soffice`/`libreoffice`) > candidats
    macOS. Retourne le chemin absolu ou None si introuvable.
    """
    settings = settings or get_settings()

    if settings.soffice_bin:
        p = Path(settings.soffice_bin)
        if p.exists():
            return str(p)

    for name in ("soffice", "libreoffice"):
        found = shutil.which(name)
        if found:
            return found

    for candidate in _MACOS_SOFFICE_CANDIDATES:
        if Path(candidate).exists():
            return candidate

    return None


def soffice_available(settings: Settings | None = None) -> bool:
    """True si LibreOffice est utilisable."""
    return find_soffice(settings) is not None


def find_textutil() -> str | None:
    """Localise `textutil` (outil natif macOS de conversion de documents).

    Permet d'extraire le texte d'un `.doc` legacy SANS LibreOffice (macOS only).
    """
    return shutil.which("textutil") or (
        "/usr/bin/textutil" if Path("/usr/bin/textutil").exists() else None
    )


def convert_doc_to_text_textutil(src: Path, *, timeout: int = 120) -> str:
    """Extrait le texte d'un `.doc`/`.rtf`/`.docx` via `textutil` (macOS).

    Raises:
        DocConverterError: textutil absent (non-macOS) ou échec de conversion.
    """
    src = Path(src)
    if not src.exists():
        raise DocConverterError(f"Fichier introuvable : {src}")
    textutil = find_textutil()
    if textutil is None:
        raise DocConverterError("textutil introuvable (disponible uniquement sur macOS).")
    try:
        proc = subprocess.run(
            [textutil, "-convert", "txt", "-stdout", str(src)],
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
        )
    except subprocess.TimeoutExpired as exc:
        raise DocConverterError(
            f"Conversion textutil expirée après {timeout}s : {src.name}"
        ) from exc
    if proc.returncode != 0:
        raise DocConverterError(
            f"Échec textutil ({src.name}) : {proc.stderr.strip()[:300]}"
        )
    return proc.stdout


def convert_doc_to_docx(
    src: Path,
    out_dir: Path | None = None,
    settings: Settings | None = None,
    *,
    timeout: int = 120,
) -> Path:
    """Convertit un `.doc` (ou autre format Word) en `.docx` via LibreOffice.

    Args:
        src: fichier source (.doc / .rtf / .odt ...).
        out_dir: dossier de sortie ; défaut = dossier temporaire.
        timeout: secondes max pour le sous-processus soffice.

    Returns:
        Chemin du `.docx` produit.

    Raises:
        LibreOfficeNotFoundError: binaire soffice introuvable.
        DocConverterError: échec de conversion / fichier de sortie absent.
    """
    src = Path(src)
    if not src.exists():
        raise DocConverterError(f"Fichier introuvable : {src}")

    soffice = find_soffice(settings)
    if soffice is None:
        raise LibreOfficeNotFoundError(
            "LibreOffice (soffice) est requis pour convertir les .doc legacy mais "
            "n'a pas été trouvé. Installez-le puis réessayez, ou renseignez "
            "SOFFICE_BIN dans .env. Voir DECISIONS.md §D1."
        )

    out_dir = Path(out_dir) if out_dir else Path(tempfile.mkdtemp(prefix="gss_doc_"))
    out_dir.mkdir(parents=True, exist_ok=True)

    # LibreOffice headless : un seul profil utilisateur jetable pour éviter les
    # conflits si une instance GUI tourne déjà.
    with tempfile.TemporaryDirectory(prefix="gss_lo_profile_") as profile:
        cmd = [
            soffice,
            "--headless",
            "--norestore",
            f"-env:UserInstallation=file://{profile}",
            "--convert-to",
            "docx",
            "--outdir",
            str(out_dir),
            str(src),
        ]
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise DocConverterError(
                f"Conversion LibreOffice expirée après {timeout}s : {src.name}"
            ) from exc

    produced = out_dir / (src.stem + ".docx")
    if proc.returncode != 0 or not produced.exists():
        raise DocConverterError(
            f"Échec conversion LibreOffice ({src.name}). "
            f"code={proc.returncode} stderr={proc.stderr.strip()[:500]}"
        )
    return produced


def load_docx_text(path: Path) -> str:
    """Extrait le texte brut d'un `.docx` (paragraphes + tables), séparé par \\n."""
    from docx import Document

    doc = Document(str(path))
    lines: list[str] = [p.text for p in doc.paragraphs]
    for table in doc.tables:
        for row in table.rows:
            cells = [c.text.strip() for c in row.cells]
            lines.append(" | ".join(cells))
    return "\n".join(lines)


def extract_pdf_text(path: Path) -> str:
    """Extrait le texte d'un PDF via PyMuPDF (pages séparées par form feed).

    Note : OCR (PDF scannés) est un fallback OPTIONNEL hors périmètre itération 1
    (OCR_ENABLED=false par défaut).
    """
    import fitz  # PyMuPDF

    parts: list[str] = []
    with fitz.open(str(path)) as doc:
        for page in doc:
            parts.append(page.get_text("text"))
    return "\f".join(parts)


def extract_doc_text(src: Path, settings: Settings | None = None):
    """Extrait le texte d'un `.doc` legacy, retourne (texte, méthode_extraction).

    Ordre de préférence :
      1. LibreOffice (conversion .docx -> python-docx) : préserve les tableaux ;
      2. textutil (natif macOS) : zéro dépendance, texte propre.

    Raises:
        LibreOfficeNotFoundError: aucune voie de conversion disponible.
    """
    from backend.schemas.common import ExtractionMethod

    src = Path(src)
    if soffice_available(settings):
        converted = convert_doc_to_docx(src, settings=settings)
        return load_docx_text(converted), ExtractionMethod.DOC_LIBREOFFICE
    if find_textutil() is not None:
        return convert_doc_to_text_textutil(src), ExtractionMethod.DOC_TEXTUTIL
    raise LibreOfficeNotFoundError(
        "Aucune voie de conversion .doc disponible : ni LibreOffice (soffice) ni "
        "textutil (macOS). Installez LibreOffice ou renseignez SOFFICE_BIN. "
        "Voir DECISIONS.md §D1."
    )


def extract_text(path: Path, settings: Settings | None = None) -> str:
    """Extrait le texte de n'importe quel format supporté (dispatch par extension).

    - `.docx` -> python-docx
    - `.doc`  -> LibreOffice si dispo, sinon textutil (macOS)
    - `.pdf`  -> PyMuPDF
    """
    path = Path(path)
    ext = path.suffix.lower()
    if ext == ".docx":
        return load_docx_text(path)
    if ext == ".doc":
        text, _ = extract_doc_text(path, settings=settings)
        return text
    if ext == ".pdf":
        return extract_pdf_text(path)
    raise DocConverterError(f"Format non supporté : {ext} ({path.name})")

File: gss-ao/backend/src/ingestion/docConverter.ts
import fs from 'fs';
import path from 'path';
import { execSync, spawnSync } from 'child_process';
import AdmZip from 'adm-zip';
import { XMLParser } from 'fast-xml-parser';
import pdf from 'pdf-parse';
import { getSettings } from '../core/config';
import { ExtractionMethod } from '../schemas/common';

export class DocConverterError extends Error {}
export class LibreOfficeNotFoundError extends DocConverterError {}

// macOS common LibreOffice paths
const MACOS_SOFFICE_CANDIDATES = [
  '/Applications/LibreOffice.app/Contents/MacOS/soffice',
  '/opt/homebrew/bin/soffice',
  '/usr/local/bin/soffice',
];

export function findSoffice(): string | null {
  const settings = getSettings();
  if (settings.sofficeBin && fs.existsSync(settings.sofficeBin)) {
    return settings.sofficeBin;
  }

  // Check PATH
  try {
    const checkCmd = process.platform === 'win32' ? 'where soffice' : 'which soffice';
    const out = execSync(checkCmd, { stdio: [] }).toString().trim();
    if (out) return out.split('\n')[0].trim();
  } catch (e) {}

  try {
    const checkCmd = process.platform === 'win32' ? 'where libreoffice' : 'which libreoffice';
    const out = execSync(checkCmd, { stdio: [] }).toString().trim();
    if (out) return out.split('\n')[0].trim();
  } catch (e) {}

  // Candidates for macOS
  if (process.platform === 'darwin') {
    for (const cand of MACOS_SOFFICE_CANDIDATES) {
      if (fs.existsSync(cand)) {
        return cand;
      }
    }
  }

  return null;
}

export function sofficeAvailable(): boolean {
  return findSoffice() !== null;
}

export function findTextutil(): string | null {
  if (process.platform !== 'darwin') return null;
  try {
    const out = execSync('which textutil', { stdio: [] }).toString().trim();
    if (out) return out;
  } catch (e) {}
  if (fs.existsSync('/usr/bin/textutil')) return '/usr/bin/textutil';
  return null;
}

export function convertDocToTextTextutil(src: string): string {
  if (!fs.existsSync(src)) {
    throw new DocConverterError(`Fichier introuvable : ${src}`);
  }
  const textutil = findTextutil();
  if (!textutil) {
    throw new DocConverterError('textutil introuvable (disponible uniquement sur macOS).');
  }
  try {
    const proc = spawnSync(textutil, ['-convert', 'txt', '-stdout', src], { timeout: 120000 });
    if (proc.status !== 0) {
      throw new DocConverterError(`Échec textutil (${path.basename(src)}) : ${proc.stderr?.toString()}`);
    }
    return proc.stdout.toString();
  } catch (e: any) {
    throw new DocConverterError(`Échec conversion textutil : ${e.message}`);
  }
}

export function convertDocToDocx(src: string, outDir?: string): string {
  if (!fs.existsSync(src)) {
    throw new DocConverterError(`Fichier introuvable : ${src}`);
  }
  const soffice = findSoffice();
  if (!soffice) {
    throw new LibreOfficeNotFoundError(
      "LibreOffice (soffice) est requis pour convertir les .doc legacy mais " +
      "n'a pas été trouvé. Installez-le puis réessayez, ou renseignez " +
      "SOFFICE_BIN dans .env."
    );
  }

  const outputDirectory = outDir || fs.mkdtempSync(path.join(path.dirname(src), 'gss_doc_'));
  const tempProfileDir = fs.mkdtempSync(path.join(path.dirname(src), 'gss_lo_profile_'));

  const args = [
    '--headless',
    '--norestore',
    `-env:UserInstallation=file:///${tempProfileDir.replace(/\\/g, '/')}`,
    '--convert-to',
    'docx',
    '--outdir',
    outputDirectory,
    src,
  ];

  try {
    const proc = spawnSync(soffice, args, { timeout: 120000 });
    if (proc.status !== 0) {
      throw new DocConverterError(
        `Échec conversion LibreOffice (${path.basename(src)}). code=${proc.status} stderr=${proc.stderr?.toString()}`
      );
    }
  } finally {
    try {
      fs.rmSync(tempProfileDir, { recursive: true, force: true });
    } catch (e) {}
  }

  const stem = path.basename(src, path.extname(src));
  const produced = path.join(outputDirectory, `${stem}.docx`);
  if (!fs.existsSync(produced)) {
    throw new DocConverterError(`Fichier converti non trouvé : ${produced}`);
  }
  return produced;
}

export interface DocxParagraph {
  text: string;
  styleName: string | null;
}

export interface DocxTable {
  rows: string[][];
}

export interface DocxStructure {
  paragraphs: DocxParagraph[];
  tables: DocxTable[];
  allElements: Array<{ type: 'paragraph'; text: string; styleName: string | null } | { type: 'table'; rows: string[][] }>;
}

export function loadDocxStructure(filePath: string): DocxStructure {
  if (!fs.existsSync(filePath)) {
    throw new DocConverterError(`Fichier non trouvé : ${filePath}`);
  }

  const zip = new AdmZip(filePath);
  const docXmlEntry = zip.getEntry('word/document.xml');
  if (!docXmlEntry) {
    throw new DocConverterError(`word/document.xml non trouvé dans le docx : ${filePath}`);
  }

  const xmlText = docXmlEntry.getData().toString('utf8');
  const parser = new XMLParser({
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
    parseAttributeValue: false,
  });

  const jsonObj = parser.parse(xmlText);
  const body = jsonObj['w:document']?.['w:body'];
  if (!body) {
    throw new DocConverterError(`Structure XML du document non reconnue dans ${filePath}`);
  }

  const paragraphs: DocxParagraph[] = [];
  const tables: DocxTable[] = [];
  const allElements: DocxStructure['allElements'] = [];

  // Helper to extract text from a paragraph node w:p
  const extractTextFromParagraphNode = (pNode: any): string => {
    if (!pNode) return '';
    let text = '';
    // w:r can be an object or an array of objects
    const runs = Array.isArray(pNode['w:r']) ? pNode['w:r'] : [pNode['w:r']];
    for (const r of runs) {
      if (!r) continue;
      // w:t can be string, object (with attributes) or array of strings/objects
      const ts = Array.isArray(r['w:t']) ? r['w:t'] : [r['w:t']];
      for (const t of ts) {
        if (!t) continue;
        if (typeof t === 'string') {
          text += t;
        } else if (typeof t === 'object') {
          if (t['#text']) {
            text += t['#text'];
          }
        }
      }
    }
    return text;
  };

  // Helper to extract style name from w:pPr -> w:pStyle
  const extractStyleFromParagraphNode = (pNode: any): string | null => {
    const pPr = pNode?.['w:pPr'];
    if (!pPr) return null;
    const pStyle = pPr['w:pStyle'];
    if (!pStyle) return null;
    return pStyle['@_w:val'] || null;
  };

  // Process elements inside body
  // The body contains w:p and w:tbl directly. fast-xml-parser might keep them as properties or in order if configured,
  // but to preserve their document order, we can parse the XML using regex to extract elements in order.
  // Wait, let's look at the child nodes of body. Since fast-xml-parser parses keys, the order of keys like w:p and w:tbl
  // might not be preserved if we just iterate them.
  // To preserve strict order of paragraphs and tables, we can scan the raw XML for <w:p> and <w:tbl> tags.
  // Let's do that! It is extremely simple to tokenize the body.
  const bodyXmlMatch = xmlText.match(/<w:body>([\s\S]*?)<\/w:body>/);
  if (!bodyXmlMatch) {
    throw new DocConverterError(`w:body non trouvé dans document.xml`);
  }
  const bodyXml = bodyXmlMatch[1];

  // Regex to match top-level tags <w:p> and <w:tbl>
  // We can write a simple parser that steps through the XML:
  let idx = 0;
  while (idx < bodyXml.length) {
    const nextP = bodyXml.indexOf('<w:p', idx);
    const nextTbl = bodyXml.indexOf('<w:tbl', idx);

    if (nextP === -1 && nextTbl === -1) {
      break;
    }

    if (nextP !== -1 && (nextTbl === -1 || nextP < nextTbl)) {
      // It's a paragraph. Find matching closing tag </w:p>
      // We must handle nested tags correctly, but w:p cannot be nested under another w:p.
      const closeP = bodyXml.indexOf('</w:p>', nextP);
      if (closeP === -1) break;
      const pXml = bodyXml.substring(nextP, closeP + 6);
      idx = closeP + 6;

      const pObj = parser.parse(pXml)['w:p'];
      if (pObj) {
        const text = extractTextFromParagraphNode(pObj);
        const styleName = extractStyleFromParagraphNode(pObj);
        const p: DocxParagraph = { text, styleName };
        paragraphs.push(p);
        allElements.push({ type: 'paragraph', text, styleName });
      }
    } else {
      // It's a table. Find matching closing tag </w:tbl>
      // w:tbl cannot be nested under another w:tbl (nested tables exist but are rare in DCE, and even if they do, simple matching works).
      const closeTbl = bodyXml.indexOf('</w:tbl>', nextTbl);
      if (closeTbl === -1) break;
      const tblXml = bodyXml.substring(nextTbl, closeTbl + 8);
      idx = closeTbl + 8;

      const tblObj = parser.parse(tblXml)['w:tbl'];
      if (tblObj) {
        const rows: string[][] = [];
        const wTrs = Array.isArray(tblObj['w:tr']) ? tblObj['w:tr'] : [tblObj['w:tr']];
        for (const tr of wTrs) {
          if (!tr) continue;
          const rowCells: string[] = [];
          const wTcs = Array.isArray(tr['w:tc']) ? tr['w:tc'] : [tr['w:tc']];
          for (const tc of wTcs) {
            if (!tc) continue;
            // A table cell tc contains paragraphs w:p
            const tcPs = Array.isArray(tc['w:p']) ? tc['w:p'] : [tc['w:p']];
            const cellTextParts: string[] = [];
            for (const p of tcPs) {
              if (p) {
                cellTextParts.push(extractTextFromParagraphNode(p));
              }
            }
            rowCells.push(cellTextParts.join(' ').trim());
          }
          rows.push(rowCells);
        }
        tables.push({ rows });
        allElements.push({ type: 'table', rows });
      }
    }
  }

  return { paragraphs, tables, allElements };
}

export function loadDocxText(filePath: string): string {
  const struct = loadDocxStructure(filePath);
  const lines: string[] = [];
  for (const el of struct.allElements) {
    if (el.type === 'paragraph') {
      lines.push(el.text);
    } else if (el.type === 'table') {
      for (const row of el.rows) {
        lines.push(row.join(' | '));
      }
    }
  }
  return lines.join('\n');
}

/**
 * Extraction ANNOTÉE d'un cadre de réponse (.docx) pour la DÉTECTION des champs à remplir.
 * Contrairement à loadDocxText (qui aplatit tout en texte), on préserve les éléments interactifs
 * du formulaire pour que l'IA repère ce qui reste à renseigner :
 *   • cases à cocher — form field `w:checkBox`, content control `w14:checkbox`, symboles Wingdings
 *     (☐ = F0A8, cochés = F0FE/F0FD/F06F/F0FB/F0FC) ou glyphes littéraux ☐ ☑ ☒ ✔ ✗
 *     → balisées « [CASE ☐ vide] » / « [CASE ☒ cochée] » ;
 *   • champs de formulaire texte (`FORMTEXT`) → « [CHAMP À REMPLIR] » ;
 *   • tableaux → rendus en grille, chaque cellule vide balisée « ⬚ » (à remplir).
 * Best-effort et tolérant (regex) : sert UNIQUEMENT à l'analyse des manques, jamais à la génération.
 */
export function loadDocxTemplateAnnotated(filePath: string): string {
  if (!fs.existsSync(filePath)) throw new DocConverterError(`Fichier non trouvé : ${filePath}`);
  const zip = new AdmZip(filePath);
  const entry = zip.getEntry('word/document.xml');
  if (!entry) throw new DocConverterError(`word/document.xml non trouvé : ${filePath}`);
  const xmlText = entry.getData().toString('utf8');
  const bodyMatch = xmlText.match(/<w:body>([\s\S]*?)<\/w:body>/);
  const bodyXml = bodyMatch ? bodyMatch[1] : xmlText;

  // Texte brut d'un fragment XML (concatène les <w:t>, décode les entités).
  const runText = (xml: string): string =>
    (xml.match(/<w:t[ >][\s\S]*?<\/w:t>/g) || [])
      .map((m) => m.replace(/<[^>]+>/g, ''))
      .join('')
      .replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"').replace(/&apos;/g, "'");

  const SYM_UNCHECKED = /w:char="F0A8"/i;                       // ☐ Wingdings
  const SYM_CHECKED = /w:char="(F0FE|F0FD|F06F|F0FB|F0FC)"/i;   // ☒/☑ Wingdings
  const LITERAL_UNCHECKED = /[☐❏❑⬜□]/; // ☐ ❏ ❑ ⬜ □
  const LITERAL_CHECKED = /[☑☒✅✔✓✗]/; // ☑ ☒ ✅ ✔ ✓ ✗

  // Analyse un fragment (paragraphe ou cellule) → ligne annotée (case à cocher / champ / texte).
  const analyzeFragment = (frag: string): string => {
    const text = runText(frag).trim();
    const hasCheckboxField = /<w:checkBox/.test(frag) || /<w14:checkbox/.test(frag);
    const checkedField = /<w:checked\b(?![^>]*w:val="0")/.test(frag) || /<w14:checked[^>]*w14:val="1"/.test(frag);
    const symUnchecked = SYM_UNCHECKED.test(frag);
    const symChecked = SYM_CHECKED.test(frag);
    const litUnchecked = LITERAL_UNCHECKED.test(text);
    const litChecked = LITERAL_CHECKED.test(text);
    const hasFormText = /FORMTEXT/.test(frag);

    if (hasCheckboxField || symUnchecked || symChecked || litUnchecked || litChecked) {
      const checked = (hasCheckboxField && checkedField) || symChecked || litChecked;
      const clean = text.replace(LITERAL_UNCHECKED, '').replace(LITERAL_CHECKED, '').replace(/\s+/g, ' ').trim();
      return `[CASE ${checked ? '☒ cochée' : '☐ vide'}] ${clean}`.trim();
    }
    if (hasFormText) return `[CHAMP À REMPLIR] ${text}`.trim();
    return text;
  };

  const lines: string[] = [];
  let idx = 0;
  while (idx < bodyXml.length) {
    const nextP = bodyXml.indexOf('<w:p', idx);
    const nextTbl = bodyXml.indexOf('<w:tbl', idx);
    if (nextP === -1 && nextTbl === -1) break;

    if (nextP !== -1 && (nextTbl === -1 || nextP < nextTbl)) {
      const closeP = bodyXml.indexOf('</w:p>', nextP);
      if (closeP === -1) break;
      const pXml = bodyXml.substring(nextP, closeP + 6);
      idx = closeP + 6;
      const line = analyzeFragment(pXml);
      if (line) lines.push(line);
    } else {
      const closeTbl = bodyXml.indexOf('</w:tbl>', nextTbl);
      if (closeTbl === -1) break;
      const tblXml = bodyXml.substring(nextTbl, closeTbl + 8);
      idx = closeTbl + 8;
      lines.push('[TABLEAU]');
      // Matrice des cellules ANALYSÉES (texte / case à cocher), puis annotation ligne↔colonne.
      const rowXmls = tblXml.match(/<w:tr[ >][\s\S]*?<\/w:tr>/g) || [];
      const matrix: string[][] = rowXmls.map((trXml) => {
        const cellXmls = trXml.match(/<w:tc[ >][\s\S]*?<\/w:tc>/g) || [];
        return cellXmls.map((tc) => analyzeFragment(tc).trim());
      });
      // En-têtes de COLONNE = 1re ligne ; libellé de LIGNE = 1re cellule non vide de la ligne.
      const colHeaders = matrix[0] || [];
      matrix.forEach((row, r) => {
        const rowHeader = (row.find((c) => c !== '') || '').replace(/^\[[^\]]+\]\s*/, '').slice(0, 60);
        const cells = row.map((cell, c) => {
          if (cell !== '') return cell;
          // Cellule VIDE à remplir : on rattache colonne + ligne pour lever l'ambiguïté.
          const col = (colHeaders[c] || '').replace(/^\[[^\]]+\]\s*/, '').slice(0, 60);
          const parts: string[] = [];
          if (r > 0 && col) parts.push(`colonne: « ${col} »`);
          if (c > 0 && rowHeader) parts.push(`ligne: « ${rowHeader} »`);
          return parts.length ? `⬚ (à remplir — ${parts.join(', ')})` : '⬚ (à remplir)';
        });
        lines.push('| ' + cells.join(' | ') + ' |');
      });
      lines.push('[/TABLEAU]');
    }
  }
  return lines.join('\n');
}

export async function extractPdfText(filePath: string): Promise<string> {
  if (!fs.existsSync(filePath)) {
    throw new DocConverterError(`Fichier non trouvé : ${filePath}`);
  }
  const dataBuffer = fs.readFileSync(filePath);
  try {
    const data = await pdf(dataBuffer);
    return data.text;
  } catch (e: any) {
    throw new DocConverterError(`Échec extraction PDF : ${e.message}`);
  }
}

export function extractDocText(src: string): { text: string; method: ExtractionMethod } {
  if (sofficeAvailable()) {
    const tempDir = fs.mkdtempSync(path.join(path.dirname(src), 'gss_lo_temp_'));
    try {
      const converted = convertDocToDocx(src, tempDir);
      const text = loadDocxText(converted);
      return { text, method: ExtractionMethod.DOC_LIBREOFFICE };
    } finally {
      try {
        fs.rmSync(tempDir, { recursive: true, force: true });
      } catch (e) {}
    }
  }
  if (findTextutil() !== null) {
    return { text: convertDocToTextTextutil(src), method: ExtractionMethod.DOC_TEXTUTIL };
  }
  throw new LibreOfficeNotFoundError(
    "Aucune voie de conversion .doc disponible : ni LibreOffice (soffice) ni " +
    "textutil (macOS). Installez LibreOffice ou renseignez SOFFICE_BIN."
  );
}

export async function extractText(filePath: string): Promise<string> {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === '.docx') {
    return loadDocxText(filePath);
  }
  if (ext === '.doc') {
    return extractDocText(filePath).text;
  }
  if (ext === '.pdf') {
    return extractPdfText(filePath);
  }
  throw new DocConverterError(`Format non supporté : ${ext} (${path.basename(filePath)})`);
}
